implementations/elixir/ockam/ockam/lib/ockam/secure_channel/channel.ex

```diff
@@ -1,380 +1,616 @@
-    key_establishment_protocol.handle_message(message, state, data)
+  defp new_xx(vault, opts) do
+    m1_payload = m2_payload = m3_payload = ""
+    {:ok, e_key_handle} = Vault.secret_generate(vault, type: :curve25519)
+    {:ok, e_public_key} = Vault.secret_publickey_get(vault, e_key_handle)
+    e = %{private: e_key_handle, public: e_public_key}
+
+    options =
+      [
+        vault: vault,
+        ephemeral_keypair: e,
+        message1_payload: m1_payload,
+        message2_payload: m2_payload,
+        message3_payload: m3_payload
+      ] ++ opts
+
+    XX.setup(options)
   end
 
-  defp handle_message(message, {:encrypted_transport, :ready} = state, data) do
-    first_address = message |> Message.onward_route() |> List.first()
+  defp vault_from_opts(encryption_options) do
+    case Keyword.fetch(encryption_options, :vault) do
+      {:ok, vault} -> {:ok, vault}
+      :error -> Ockam.Vault.Software.init()
+    end
+  end
 
-    cond do
-      first_address === data.ciphertext_address ->
-        case decrypt_and_send_to_router(message, state, data) do
-          {:next_state, state, data} ->
-            {:next_state, state, data}
+  defp identity_from_opts(options) do
+    identity_module =
+      Keyword.get_lazy(options, :identity_module, &Identity.default_implementation/0)
 
-          {:error, reason} ->
-            Logger.warn("packet dropped: #{inspect(reason)}")
-            # Erroneous/Repeated/Malicius packet dropped.
-            # decryptor state is not affected
-            {:next_state, state, data}
+    case Keyword.fetch(options, :identity) do
+      {:ok, :dynamic} ->
+        with {:ok, identity, _id} <- Identity.create(identity_module) do
+          {:ok, identity}
         end
 
-      first_address === data.plaintext_address ->
-        encrypt_and_send_to_peer(message, state, data)
+      {:ok, other} ->
+        Identity.make_identity(identity_module, other)
 
-      true ->
-        {:next_state, state, data}
+      :error ->
+        {:error, :missing_identity}
     end
   end
 
-  defp decode_payload(payload) do
-    case :bare.decode(payload, :data) do
-      {:ok, encrypted, ""} -> {:ok, encrypted}
-      _other -> {:error, {:bad_payload, payload}}
+  def inner_setup_impl(address, inner_address, options) do
+    trust_policies = Keyword.get(options, :trust_policies, [])
+    additional_metadata = Keyword.get(options, :additional_metadata, %{})
+    encryption_options = Keyword.get(options, :encryption_options, [])
+    key_exchange_timeout = Keyword.get(options, :key_exchange_timeout, @handshake_timeout)
+    vault_name = Keyword.get(options, :vault_name)
+
+    with {:ok, role} <- Keyword.fetch(options, :role),
+         {:ok, vault} <- vault_from_opts(encryption_options),
+         {:ok, identity} <- identity_from_opts(options),
+         {:ok, key_exchange_state} <-
+           new_xx(vault, Keyword.take(encryption_options, [:static_keypair])) do
```

Comment:
I know the module is called `XX`, but can we use more descriptive function name here?